ts/parts-3d/Column.ts

```diff
@@ -533,18 +543,42 @@ wrap(H.StackItem.prototype, 'getStackBox', function (
     chart: Highcharts.Chart
 ): void { // #3946
     var stackBox = proceed.apply(this, [].slice.call(arguments, 1));
-
-    // Only do this for 3D chart.
+    // Only do this for 3D column and iherited series.
     if (chart.is3d()) {
-        var pos = ({
-            x: stackBox.x,
-            y: stackBox.y,
-            z: 0
+        var columnSeries: Highcharts.ColumnSeries | undefined;
+        // Check if any series iherits from column 3D
+        chart.series.forEach(function (
+            series: Highcharts.Series
+        ): void {
+            if (series instanceof seriesTypes.column && series.data.length) {
+                columnSeries = series;
+            }
         });
+        // If any series is a column series, use its barW, z and depth
+        // parameters for correct stackLabels position calculation
+        if (columnSeries) {
+            var height: number = arguments[6],
+                pos = ({
+                    x: stackBox.x +
+                        (
+                            chart.inverted ?
+                                height :
+                                (columnSeries as any).barW / 2
+                        ),
+                    y: stackBox.y,
+                    z: columnSeries.z + (columnSeries.options as any).depth / 2
+                });
 
-        pos = H.perspective([pos], chart, true)[0];
-        stackBox.x = pos.x;
-        stackBox.y = pos.y;
+            if (chart.inverted) {
+                // Do not use default offset calculation logic for 3D
+                // inverted stackLabels.
+                stackBox.width = 0;
+            }
+
+            pos = H.perspective([pos], chart, true, false)[0];
```

Comment:
How about renaming `pos` to something more descriptive? and adding comments when the value of the pos variable is changing significantly (from flat to 3d, from not adjusted to adjusted, etc.), or maybe even using more variables for better readability (and TS types might be easier to write).